fireplace/cards/utils.py

```diff
@@ -38,6 +38,50 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
+# take a list of weights and a list of card pools and produce a random weighted sample without replacement
+def weighted_card_choice(source, weights, card_sets, count):
+		# get all the cards matching each set of filters and create the weight line
+		# (user does not have to normalize weighting from 0-1)
+		cardlist = []
+		weightedsegmentstarts = [0] * len(card_sets)
+		weightlinelength = 0
+
+		for i, c in enumerate(card_sets):
+			weightedsegmentstarts[i] = len(cardlist)
+			cardlist += c
+			# higher weighted sets (segments) occupy more space per item on the weight line
+			weightlinelength += len(c) * weights[i]
+
+		ret = []
+
+		# for each card
+		for c in range(count):
+
+			# pick a point along the weight line
+			r = random.random() * weightlinelength
+			pos = 0
+
+			# figure out which segment it's in
+			for i, w in enumerate(weights):
+				segmentarraylength = ((len(cardlist) if i == len(weights)-1 else weightedsegmentstarts[i+1]) - weightedsegmentstarts[i])
+				segmentlinelength = segmentarraylength * w
+				# it's in this segment, find exact card
+				if (r < pos + segmentlinelength):
+					# find percentage along this segment we're at
+					spc = (r - pos) / segmentlinelength
```

Comment:
It's not clear to me whether you're intending to do floored integral division, or whether you really want floats (here and other places). Could you clarify?